src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,122 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int INITIAL_CAPACITY = 16;
+    private static final double LOAD_FACTOR = 0.75;
+    private static final int CAPACITY_MULTIPLIER = 2;
+    private Node<K, V>[] table;
+    private int size;
+    private int threshold;
+
+    public MyHashMap() {
+        table = new Node[INITIAL_CAPACITY];
+        size = 0;
+        threshold = (int) (LOAD_FACTOR * INITIAL_CAPACITY);
+    }
 
     @Override
     public void put(K key, V value) {
-
+        resizeCapacity();
+        Node<K, V> node = new Node<>(key, value);
+        int index = indexOf(node.hash);
+        if (table[index] == null) {
+            createBucket(index, node);
+        } else {
+            updateBucket(index, node);
+        }
     }
 
     @Override
     public V getValue(K key) {
+        int keyHash = hash(key);
+        int index = indexOf(keyHash);
+        if (size == 0 || table[index] == null) {
+            return null;
+        }
+        Node<K, V> bucket = table[index];
+        if (bucket.hash == keyHash && objectsAreEqual(bucket.key, key)) {
+            return bucket.value;
+        }
+        while (bucket != null) {
+            if (bucket.hash == keyHash && objectsAreEqual(bucket.key, key)) {
+                return bucket.value;
+            }
+            bucket = bucket.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private int indexOf(int hash) {
+        return Math.abs(hash % table.length);
+    }
+
+    private void resizeCapacity() {
+        if (size < threshold) {
+            return;
+        }
+        Node<K, V>[] oldTable = table;
+        table = new Node[oldTable.length * CAPACITY_MULTIPLIER];
+        int oldSize = size;
+        size = 0;
+        threshold *= CAPACITY_MULTIPLIER;
+        transferDataToNewEmptyTable(oldTable, oldSize);
+    }
+
+    private boolean objectsAreEqual(K a, K b) {
+        return a == b || a != null && a.equals(b);
+    }
```

Comment:
I would like use Objects.equals() but I'm still not sure if we are allowed to use it already, since in the previous homeworks we were restricted to do so.